Credits for 8.10 (#9939)
coq/coq: doc/sphinx/changes.rst

Recent changes
--------------

Version 8.10
------------

Summary of changes
~~~~~~~~~~~~~~~~~~

|Coq| version 8.10 contains two major new features: support for a native
fixed-precision integer type and a new sort `SProp` of strict
propositions. It is also the result of refinements and stabilization of
features, deprecations or removals of deprecated features, cleanups of
the internals of the system and API. This release includes many
user-visible changes, including deprecations that are documented in

> ``CHANGES.md`` and the following section, and new features that are

Author: Indeed, I realized afterwards, fixed

    Laporte.

  - The `SProp` sort of definitionally proof-irrelevant propositions was
    introduced by Gaëtan Gilbert. `SProp` allows to mark proof terms as
    irrelevant for conversion, and is treated like `Prop` during
    extraction. It is enabled using `-allow-sprop`.

  - The unfolding heuristic in termination checking was made more complete
    by Enrico Tassi, allowing more constants to be unfolded to discover
    valid recursive calls.

- Notations:

  - New command `String Notation` to register string syntax for custom
    inductive types, by Jason Gross.

  - Numeral Notations now parse decimal constants, by Pierre Roux.

- Ltac:

  - Ltac backtraces can be turned on using the "Ltac Backtrace" option,
    by Pierre-Marie Pédrot.

- Universes:

  - Added `Print Universes Subgraph` variant of `Print Universes` and
    private universes for opaque polymorphic constants, by Gaëtan Gilbert.

- The tactics 'lia','nia','lra','nra' are now using a novel
  Simplex-based proof engine, by Fréderic Besson.

- `Combined Scheme` now works when inductive schemes are generated in sort
  `Type`, by Théo Winterhalter.

- CoqIDE:

>   - Migrated to gtk+3 and lablgtk3 by Hugo Herbelin.

Reviewer: Yes, I'd credit @garrigue here as well.

>   - Migrated to gtk+3 and lablgtk3 by Hugo Herbelin.

Author: I don't see how to phrase it, but I've added @garrigue here.

  - Supports smart input for Unicode characters by Arthur Charguéraud.

- SSReflect has new intro patterns and a consistent clear discipline, by

>   Enrico Tassi.

Author: @gares how would you say it?

- Build system:

>   - The project is migrating to `dune`, by Emilio Jesús Gallego Arias.

Reviewer 2: Thanks @Zimmi48 , I'd write "which is scheduled for deprecation" instead of the removal. Also, worth adding could be "the Dune build system provides an improved experience for plugin developers".

>   - The project is migrating to `dune`, by Emilio Jesús Gallego Arias.

Reviewer 2: We should also credit @rgrinberg here and thank the Dune team.

Version 8.10 also comes with a bunch of smaller-scale changes and
improvements regarding the different components of the system, including
many additions to the standard library.

> 

Reviewer 2: Worth mentioning on the main list of changes is that the CI system now tests Coq against OCaml trunk. This is IMO worth to be advertised more widely to OCaml people.

Melquiond, Matthieu Sozeau, Enrico Tassi (who migrated it to opam 2)
with contributions from many users. A list of packages is available at
https://coq.inria.fr/opam/www/.

The 61 contributors to this version are David A. Dalrymple, Tanaka
Akira, Benjamin Barenblat, Yves Bertot, Frédéric Besson, Lasse
Blaauwbroek, Martin Bodin, Joachim Breitner, Tej Chajed, Frédéric
Chapoton, Arthur Charguéraud, Cyril Cohen, Lukasz Czajka, Christian
Doczkal, Maxime Dénès, Andres Erbsen, Jim Fehrle, Gaëtan Gilbert, Matěj
Grabovský, Simon Gregersen, Jason Gross, Samuel Gruetter, Hugo Herbelin,
Jasper Hugunin, Mirai Ikebuchi, Emilio Jesus Gallego Arias, Chantal
Keller, Matej Košík, Vincent Laporte, Olivier Laurent, Larry Darryl Lee
Jr, Pierre Letouzey, Nick Lewycky, Yao Li, Yishuai Li, Xia Li-yao, Assia
Mahboubi, Simon Marechal, Erik Martin-Dorel, Thierry Martinez, Guillaume
Melquiond, Kayla Ngan, Sam Pablo Kuper, Karl Palmskog, Clément
Pit-Claudel, Pierre-Marie Pédrot, Pierre Roux, Kazuhiko Sakaguchi, Ryan
Scott, Vincent Semeria, Gan Shen, Michael Soegtrop, Matthieu Sozeau,
Enrico Tassi, Laurent Théry, Kamil Trzciński, whitequark, Théo

> Winterhalter, Beta Ziliani and Théo Zimmermann.

Author: I used basically the command shown by Emilio

> Winterhalter, Beta Ziliani and Théo Zimmermann.

Reviewer 3: OK. Let's try to find a way to enhance this command to include co-authors.

Many power users helped to improve the design of the new features via
the issue and pull request system, the |Coq| development mailing list or
the coq-club@inria.fr mailing list. It would be impossible to mention
exhaustively the names of everybody who to some extent influenced the
development.

Version 8.10 is the fifth release of |Coq| developed on a time-based
development cycle. Its development spanned 6 months from the release of
|Coq| 8.9. Vincent Laport is the release manager and maintainer of this
release. This release is the result of ??? commits and ??? PRs
merged, closing ??? issues.

| Santiago de Chile, April 2019,
| Matthieu Sozeau for the |Coq| development team
|

> Details of changes

Author: I'd keep the detailed changes instead to avoid scattering info, even if it costs a bit of duplication.

> Details of changes

Reviewer 3: OK.

~~~~~~~~~~~~~~~~~~

OCaml and dependencies

- Coq 8.10 requires OCaml >= 4.05.0, bumped from 4.02.3 See the

>   INSTALL file for more information on dependencies.

Reviewer 2: This should definitively be bumped to the main changes section.

- Coq 8.10 doesn't need Camlp5 to build anymore. It now includes a
  fork of the core parsing library that Coq uses, which is a small
  subset of the whole Camlp5 distribution. In particular, this subset
  doesn't depend on the OCaml AST, allowing easier compilation and

>   testing on experimental OCaml versions.

Author: Don't have time to do it now, feel free to add credits to the detailed changes

  The Coq developers would like to thank Daniel de Rauglaudre for many
  years of continued support.

Coqide

> - CoqIDE now depends on gtk+3 and lablgtk3, rather than gtk+2 and lablgtk2.

Reviewer: This is already said before.

Reviewer 3: Yes, that's the whole question of “Summary of changes” + “Detailed changes” or “Main changes” + “Other changes”. @mattam82 expressed how he preferred a consistent view in “Detailed changes” at the price of duplication (that's what we had before with the `CHANGES` file).

- CoqIDE now properly sets the module name for a given file based on
  its path, see -topfile change entry for more details.

Coqtop

- the use of `coqtop` as a compiler has been deprecated, in favor of
  `coqc`. Consequently option `-compile` will stop to be accepted in
  the next release. `coqtop` is now reserved to interactive

>   use. (@ejgallego #9095)

Reviewer 2: I'm all for it, and also solves the problem of crediting reviewers as it just one click away.

>   use. (@ejgallego #9095)

Author: Unless you have time to do it Theo I would rather suggest to make it a requirement for entries in changes in the future.

  Projections from non-primitive (emulated) records are allowed with
  warning "nonprimitive-projection-syntax".

Kernel

- Added primitive integers

- Unfolding heuristic in termination checking made more complete.
  In particular Coq is now more aggressive in unfolding constants
  when it looks for a iota redex. Performance regression may occur
  in Fixpoint declarations without an explicit {struct} annotation,
  since guessing the decreasing argument can now be more expensive.
  (PR #9602)

Notations

- New command `Declare Scope` to explicitly declare a scope name
  before any use of it. Implicit declaration of a scope at the time of
  `Bind Scope`, `Delimit Scope`, `Undelimit Scope`, or `Notation` is
  deprecated.

- New command `String Notation` to register string syntax for custom
  inductive types.

- Numeral notations now parse decimal constants such as 1.02e+01 or
  10.2. Parsers added for Q and R. This should be considered as an
  experimental feature currently.
  Note: in -- the rare -- case when such numeral notations were used
  in a development along with Q or R, they may have to be removed or
  deconflicted through explicit scope annotations (1.23%Q,

>   45.6%R,...).

Reviewer 4: The feature has just been ported to 8.10.

- Various bugs have been fixed (e.g. PR #9214 on removing spurious
  parentheses on abbreviations shortening a strict prefix of an application).

- Numeral Notations now support inductive types in the input to
  printing functions (e.g., numeral notations can be defined for terms
  containing things like `@cons nat O O`), and parsing functions now
  fully normalize terms including parameters of constructors (so that,
  e.g., a numeral notation whose parsing function outputs a proof of
  `Nat.gcd x y = 1` will no longer fail to parse due to containing the
  constant `Nat.gcd` in the parameter-argument of `eq_refl`).  See
  #9840 for more details.

- Deprecated compatibility notations have actually been removed.  Uses
  of these notations are generally easy to fix thanks to the hint

>   contained in the deprecation warnings. For projects that require

Reviewer 3: ```suggestion
  contained in the deprecation warning emitted by the previous version of Coq. For projects that require
```

  more than a handful of such fixes, there is [a
  script](https://gist.github.com/JasonGross/9770653967de3679d131c59d42de6d17#file-replace-notations-py)
  that will do it automatically, using the output of coqc. The script
  contains documentation on its usage in a comment at the top.

Plugins

> - The quote plugin (https://coq.inria.fr/distrib/V8.8.1/refman/proof-engine/detailed-tactic-examples.html#quote)

Reviewer 3: This link could now be: https://coq.inria.fr/distrib/V8.9.0/refman/proof-engine/detailed-tactic-examples.html#quote

Author: Right

  was removed. If some users are interested in maintaining this plugin
  externally, the Coq development team can provide assistance for extracting
  the plugin and setting up a new repository.

Tactics

- Removed the deprecated `romega` tactics.
- Tactic names are no longer allowed to clash, even if they are not defined in
  the same section. For example, the following is no longer accepted:

>   `Ltac foo := idtac. Section S. Ltac foo := fail. End S.`

Reviewer 3: I think this is more appropriate in an `Ltac` category than in `Tactics`...

- The tactics 'lia','nia','lra','nra' are now using a novel
  Simplex-based proof engine. In case of regression, 'Unset Simplex'

>   to get the venerable Fourier-based engine.

Reviewer 3: If we decide to have a "Main changes" / "Other changes" separation, then the note about `Unset Simplex` should move to the previous section.

- Names of existential variables occurring in Ltac functions
  (e.g. `?[n]` or `?n` in terms - not in patterns) are now interpreted
  the same way as other variable names occurring in Ltac functions.

- Hint declaration and removal should now specify a database (e.g. `Hint Resolve
  foo : database`). When the database name is omitted, the hint is added to the

>   core database (as previously), but a deprecation warning is emitted.

Reviewer 3: ```suggestion
  `core` database (as previously), but a deprecation warning is emitted.
```

- There are now tactics in `PreOmega.v` called
  `Z.div_mod_to_equations`, `Z.quot_rem_to_equations`, and
  `Z.to_euclidean_division_equations` (which combines the `div_mod`
  and `quot_rem` variants) which allow `lia`, `nia`, `romega`, etc to
  support `Z.div` and `Z.modulo` (`Z.quot` and `Z.rem`, respectively),
  by posing the specifying equation for `Z.div` and `Z.modulo` before
  replacing them with atoms.

- Ltac backtraces can be turned on using the "Ltac Backtrace" option.

> - The syntax of the `autoapply` tactic was fixed to conform with preexisting

Reviewer 3: ```suggestion
- The syntax of the :tacn:`autoapply` tactic was fixed to conform with preexisting
```

  projection and application nodes (bug #9508).

- `Hypotheses` and `Variables` can now take implicit binders inside sections.

- Removed deprecated option `Automatic Coercions Import`.

- The `Show Script` command has been deprecated.

- Option `Refine Instance Mode` has been deprecated and will be removed in
  the next version.

- `Coercion` does not warn ambiguous paths which are obviously convertible with
  existing ones.

- A new flag `Fast Name Printing` has been introduced. It changes the
  algorithm used for allocating bound variable names for a faster but less
  clever one.

Tools

- The `-native-compiler` flag of `coqc` and `coqtop` now takes an argument which can have three values:
  - `no` disables native_compute
  - `yes` enables native_compute and precompiles `.v` files to native code
  - `ondemand` enables native_compute but compiles code only when `native_compute` is called

  The default value is `ondemand`.

  Note that this flag now has priority over the configure flag of the same name.

- A new `-bytecode-compiler` flag for `coqc` and `coqtop` controls whether
  conversion can use the VM. The default value is `yes`.

> - CoqIDE now supports input for Unicode characters. For example, typing

Author: I let you reorganize as you see fit, I can’t do complicated edits from my iPad here :)

  "\alpha" then the "Shift+Space" will insert the greek letter alpha.
  In fact, typing the prefix string "\a" is sufficient.
  A larger number of default bindings are provided, following the latex
  naming convention. Bindings can be customized, either globally, or on a
  per-project basis, with the requirement is that keys must begin with a
  backslash and contain no space character. Bindings may be assigned custom
  priorities, so that prefixes resolve to the most convenient bindings.

>   The documentation pages for CoqIDE provides further details.

Author: Right

- The prelude used to be automatically Exported and is now only
  Imported. This should be relevant only when importing files which
  don't use -noinit into files which do.

- Added `Coq.Structures.OrderedTypeEx.String_as_OT` to make strings an
  ordered type (using lexical order).

- The `Coq.Numbers.Cyclic.Int31` library is deprecated.

- Added lemmas about `Z.testbit`, `Z.ones`, and `Z.modulo`.

- Moved the `auto` hints of the `FSet` library into a new
  `fset` database.

Universes

- Added `Print Universes Subgraph` variant of `Print Universes`.
  Try for instance `Print Universes Subgraph(sigT2.u1 sigT_of_sigT2.u1 projT3_eq.u1 eq_sigT2_rect.u1).`

- Added private universes for opaque polymorphic constants, see doc
  for the "Private Polymorphic Universes" option (and Unset it to get
  the previous behaviour).

SProp

- Added a universe "SProp" for definitionally proof irrelevant
  propositions. Use with -allow-sprop. See manual for details.

Inductives

- An option and attributes to control the automatic decision to
  declare an inductive type as template polymorphic were added.
  Warning "auto-template" will trigger when an inductive is

>   automatically declared template polymorphic without the attribute.

Author: Indeed

Funind

- Inductive types declared by Funind will never be template polymorphic.

Misc

- Option "Typeclasses Axioms Are Instances" is deprecated. Use Declare Instance for axioms which should be instances.

- Removed option "Printing Primitive Projection Compatibility"

SSReflect

- New intro patterns:
  - temporary introduction: `=> +`
  - block introduction: `=> [^ prefix ] [^~ suffix ]`
  - fast introduction: `=> >`
  - tactics as views: `=> /ltac:mytac`
  - replace hypothesis: `=> {}H`

>   See the reference manual for the actual documentation.

Reviewer: What do you mean Theo? (We can't say, see the doc section 4.2, since numbers are gone)

>   See the reference manual for the actual documentation.

Reviewer 3: We can do:

```rst
See Section :ref:`introduction-in-the-context`.
```

Grabovský, Simon Gregersen, Jason Gross, Samuel Gruetter, Hugo Herbelin,
Jasper Hugunin, Mirai Ikebuchi, Emilio Jesus Gallego Arias, Chantal
Keller, Matej Košík, Vincent Laporte, Olivier Laurent, Larry Darryl Lee
Jr, Pierre Letouzey, Nick Lewycky, Yao Li, Yishuai Li, Xia Li-yao, Assia
Mahboubi, Simon Marechal, Erik Martin-Dorel, Thierry Martinez, Guillaume
Melquiond, Kayla Ngan, Sam Pablo Kuper, Karl Palmskog, Clément
Pit-Claudel, Pierre-Marie Pédrot, Pierre Roux, Kazuhiko Sakaguchi, Ryan
Scott, Vincent Semeria, Gan Shen, Michael Soegtrop, Matthieu Sozeau,
Enrico Tassi, Laurent Théry, Kamil Trzciński, whitequark, Théo
Winterhalter, Beta Ziliani and Théo Zimmermann.

Many power users helped to improve the design of the new features via
the issue and pull request system, the |Coq| development mailing list,
the coq-club@inria.fr mailing list or the new Discourse forum. It would
be impossible to mention exhaustively the names of everybody who to some
extent influenced the development.

Version 8.10 is the fifth release of |Coq| developed on a time-based
development cycle. Its development spanned 6 months from the release of

> |Coq| 8.9. Vincent Laport is the release manager and maintainer of this

Reviewer 3: ```suggestion
|Coq| 8.9. Vincent Laporte is the release manager and maintainer of this
```
